prov/verbs/src/verbs_ep.c

```diff
@@ -405,6 +407,41 @@ vrb_alloc_init_ep(struct fi_info *info, struct vrb_domain *domain,
 	return NULL;
 }
 
+/* Generate flush completion entries for any queued send requests.
+ * We only need to record the wr_id and that the entry was not a
+ * receive (indicated by lack of IBV_WC_RECV flag).
+ */
+static void vrb_flush_sq(struct vrb_ep *ep)
+{
+	struct vrb_context *ctx;
+	struct vrb_cq *cq;
+	struct slist_entry *entry;
+	struct ibv_wc wc = {0};
+
+	if (!ep->util_ep.tx_cq)
+		return;
+
+	cq = container_of(ep->util_ep.tx_cq, struct vrb_cq, util_cq);
+	wc.status = IBV_WC_WR_FLUSH_ERR;
+	wc.vendor_err = FI_ECANCELED;
+
+	cq->util_cq.cq_mutex_lock(&cq->util_cq.cq_lock);
+	while (!slist_empty(&ep->sq_list)) {
+		entry = slist_remove_head(&ep->sq_list);
+		ctx = container_of(entry, struct vrb_context, entry);
+		assert(ctx->op_ctx == VRB_POST_SQ);
+
+		wc.wr_id = (uintptr_t) ctx->user_ctx;
+		cq->credits++;
+		ctx->ep->sq_credits++;
+		ofi_buf_free(ctx);
```

Comment:
If ctx is freed at here, but the completion event is eventually returned by the later call of ibv_poll_cq(), then the  ctx can be already reused and vrb_poll_cq() can check wrong thing correct?

Reply:
let's see sean's comment, in my understanding the ctx memory seems will not be used any more, here it just write an event (vrb_save_wc) and later call of ibv_poll_cq() need not do anything for that event any more.

Reply:
ok, a problem is if it freed here, if it is possible (in any case), the same event being pop-up at path of vrb_poll_cq()->ibv_poll_cq()?

Reply:
The ctx memory is not accessed beyond the free call.  The saved work completion (wc.wr_id) references ctx->user_ctx, which is what gets returned.

The higher-level flow is this on the send path:
```
wr.wr_id = user_ctx
vrb_post_send(wr)
/* swap in our context immediately before posting */
ctx = alloc_ctx()
ctx->user_ctx = wr.wr_id
wr.wr_id = ctx
ibv_post_send(wr)
```
The completion handling does the opposite:
```
/* swap back to the user's context */
ctx = wc.wr_id
wc.wr_id = ctx->user_ctx
free(ctx)
report completion to user...
```
The reason for the context swapping is that verbs does not return certain information in the work completion (like if the operation was a send or receive) if the operation failed.  The ctx structure used here stores that information, so that we have it in order to recover from errors.

Reply:
I chatted with Xuezhao, so there is no CQ anymore (no polling) after flush correct?

Reply:
vrb_flush_sq() only being called inside vrb_ep_close() -
vrb_ep_close() {
     ...
     vrb_cleanup_cq();
     vrb_flush_sq();
     vrb_close_free_ep();
}
So seems yes, after vrb_flush_sq() it will never pool the CQ any more.

This real problem looks like is a IB verbs bug, because the transmit OP submitted to IB QP (vrb_post_send -> ibv_post_send), but IB verbs never reports its completion by ibv_poll_cq().
This patch in OFI verbs provider workaround it by adding per EP tracking queue (sq_list), I think it is safer than workaround it in mercury.

Reply:
Technically, the CQ is still there, but the QP has been freed (shown above in the '...' portion of the code).  After freeing the QP, we flush the CQ of all completions.  Only then do we generate completions for any unmatched sends.

There should only be an issue if the verbs HW suddenly starts generating completions for QPs that no longer exist.